Fix screenshot crash with long video title, #3334

low-batt opened the pull request:
This commit will:

- Change the method  MPVController.handleEvent to raise an alert if
    screenshot generation failed
- Change the method  Utility.getLatestScreenshot to guard against an empty
    screenshot directory

This commit adds new text for the alert that will require localization.

- [ ] This change has been discussed with the author.
- [ x] It implements / fixes issue #3334.

---

**Description:**

NOTE the need for localization of the added alert message text.

low-batt committed:
Fix screenshot crash with long video title, iina#3334

This commit will:

- Change the method  MPVController.handleEvent to raise an alert if
    screenshot generation failed
- Change the method  Utility.getLatestScreenshot to guard against an empty
    screenshot directory

This commit adds new text for the alert that will require localization.

saagarjha commented on iina/MPVController.swift:
```diff
@@ -664,6 +664,18 @@ class MPVController: NSObject {
       let reply = event.pointee.reply_userdata
       if reply == MPVController.UserData.screenshot {
+        let code = event.pointee.error
+        guard 0 <= code else {
```
I think I would prefer the condition to be flipped

```suggestion
        guard code >= 0 else {
```

low-batt replied:
Fixed

saagarjha commented on iina/MPVController.swift:
```diff
@@ -664,6 +664,18 @@ class MPVController: NSObject {
       if reply == MPVController.UserData.screenshot {
+        let code = event.pointee.error
+        guard 0 <= code else {
+          let error = String.init(cString: mpv_error_string(code))
```
```suggestion
          let error = String(cString: mpv_error_string(code))
```

low-batt replied:
I see my inexperience with Swift is showing. Fixed.

saagarjha commented on iina/Utility.swift:
```diff
@@ -461,7 +461,7 @@ class Utility {
       at: folder,
       includingPropertiesForKeys: [.creationDateKey],
-      options: .skipsSubdirectoryDescendants) else { return nil }
+      options: .skipsSubdirectoryDescendants), !contents.isEmpty else { return nil }
```
```suggestion
      options: .skipsSubdirectoryDescendants),
      let latestFile = contents.first else { return nil }
```

low-batt replied:
Fixed. I did have to use var instead of let as the code below this updates latestFile.

saagarjha replied:
Ah, I see. I would probably replace the code below to use `min(by:)` with a custom comparator, to be honest, but it's up to you if you want to change that here.

low-batt replied:
That was existing code that I thought could be simplified, but I chose to make the minimum change to fix the issue. I have just pushed a commit that simplifies this method. See if that is what you were expecting.

low-batt committed:
Fix screenshot crash with long video title, #3334

This commit will address review comments on the changes to the methods
MPVController.handleEvent and Utility.getLatestScreenshot.

low-batt committed:
Fix screenshot crash with long video title, #3334

This commit will:
- Extend URL with a new optional computed property, creationDate
- Change the method Utility.getLatestScreenshot to use the array max method
  to find the latest screenshot file

uiryuu commented on iina/Utility.swift:
```diff
@@ -461,18 +461,9 @@ class Utility {
-    return latestFile
+      options: .skipsSubdirectoryDescendants),
+          !contents.isEmpty else { return nil }
+    return contents.max { a, b in a.creationDate! < b.creationDate! }
```
```suggestion
    return contents.filter { $0.creationDate != nil }.max { $0.creationDate! < $1.creationDate! }
```
Or if you are sure that `creationDate` would never be `nil`, you can remove the `filter` clause.

low-batt replied:
Fixed. Rebased with develop. Tested again. I kept the filter. Made me worry that someone could put the screen shot directory on some network device that doesn't provide it.

low-batt committed:
Fix screenshot crash with long video title, #3334

Rebased with develop and addressed review comment.

This commit will:
- Extend URL with a new optional computed property, creationDate
- Change the method Utility.getLatestScreenshot to use the array max method
  to find the latest screenshot file
